modal/shared/logging.py

```diff
@@ -0,0 +1,107 @@
+from datadog_api_client.v2.model.http_log_item import HTTPLogItem
+
+
+def get_logger(name: str):
+    return logging.getLogger(name)
+
+
+# Define a custom logging handler that sends logs to Datadog
+class DatadogHandler(logging.Handler):
+    def __init__(self, api_client: ApiClient):
+        self.api_client = api_client
+
+        super().__init__()
+
+    def emit(self, record):
+        # Ignore debug messages
+        if record.levelno == logging.DEBUG:
+            return
+
+        environment_name = os.getenv("DD_ENV") or "development"
+
+        log_payload = {
+            "python-logging": {
+                "py-env": "development",
+                "py-message": record.getMessage(),
+                "py-status": record.levelname.lower(),
+                "py-logger": record.name,
+                "py-stacktrace": record.exc_info,
+                "py-exception": record.exc_text,
+                "py-line": record.lineno,
+                "py-file": record.filename,
+                "py-function": record.funcName,
+                "py-level": record.levelno,
+                "py-path": record.pathname,
+                "py-thread": record.thread,
+                "py-threadName": record.threadName,
+                "py-process": record.process,
+                "py-processName": record.processName,
+                "py-args": record.args,
+                "py-msecs": record.msecs,
+                "py-relativeCreated": record.relativeCreated,
+                "py-created": record.created,
+                "py-module": record.module,
+            },
+            "message": record.getMessage(),
+            "environment": environment_name,
+            "source": record.name,
+            "model": record.__dict__.get("model"),
+            "modal": {
+                "cloudProvider": os.getenv("MODAL_CLOUD_PROVIDER") or "",
+                "environment": os.getenv("MODAL_ENVIRONMENT") or "",
+                "imageId": os.getenv("MODAL_IMAGE_ID") or "",
+                "region": os.getenv("MODAL_REGION") or "",
+                "taskId": os.getenv("MODAL_TASK_ID") or "",
+            },
+        }
+
+        # Send the log to Datadog using the Logs API
+        try:
+            body = HTTPLog(
+                [
+                    HTTPLogItem(
+                        ddsource="Python",
+                        ddtags=f"env:{environment_name}",
+                        hostname=os.getenv("MODAL_TASK_ID") or "",
+                        message=json.dumps(log_payload),
+                        service="openrouter-runner",
+                        status=record.levelname.lower(),
+                    ),
+                ]
+            )
+
+            logs = LogsApi(api_client)
+
+            logs.submit_log(
+                content_encoding=ContentEncoding.DEFLATE,
+                body=body,  # type: ignore
+            )
+
+        except Exception as e:
+            print(f"Error sending log to Datadog: {e}")
```

Comment:
think we should still do a bare `logging.exception(` in order for the sentry hook to fire & notify us there?

```suggestion
            logging.exception("Error sending log to Datadog")
```

Reply:
Added a manual sentry capture here so we don't create a error loop trying to send the log to DD again